src/physlr-molecules.cc

```diff
@@ -0,0 +1,398 @@
+				g[u].name = node1;
+				g[u].weight = weight;
+				g[u].indexOriginal = u;
+				barcodeToIndex[node1] = u;
+				indexToBarcode[u] = node1;
+			} else {
+				ss >> node2 >> weight;
+				E = boost::add_edge(barcodeToIndex[node1], barcodeToIndex[node2], g).first;
+				g[E].weight = weight;
+			}
+		}
+	}
+	if (verbose) {
+		std::cerr << "Loaded edges to graph ";
+	} else {
+		std::cerr << "Loaded graph ";
+	}
+#if _OPENMP
+	std::cerr << "in sec: " << omp_get_wtime() - sTime << std::endl;
+	sTime = omp_get_wtime();
+#endif
+	std::cerr << "Memory usage: " << double(memory_usage()) / double(1048576) << "GB" << std::endl;
+}
+
+// Generate a new graph from an old graph using component information from molecule separation
+void
+componentsToNewGraph(
+    const graph_t& oldG,
+    graph_t& newG,
+    vecVertexToComponent_t& vecVertexToComponent)
+{
+	barcodeToIndex_t newGBarcodeToIndex;
+#if _OPENMP
+	double sTime = omp_get_wtime();
+#endif
+	for (size_t i = 0; i < vecVertexToComponent.size(); i++) {
+
+		size_t maxVal = 0;
+		for (auto&& val : vecVertexToComponent[i]) {
+			if (val.second > maxVal) {
+				maxVal = val.second;
+			}
+		}
+
+		for (size_t j = 0; j < maxVal + 1; j++) {
+			vertex_t u = boost::add_vertex(newG);
+			newG[u].name = oldG[i].name + "_" + std::to_string(j);
+			newG[u].weight = oldG[i].weight;
+			newG[u].indexOriginal = u;
+			newGBarcodeToIndex[newG[u].name] = u;
+		}
+	}
+
+	auto edgeItRange = boost::edges(oldG);
+	for (auto edgeIt = edgeItRange.first; edgeIt != edgeItRange.second; ++edgeIt) {
+		vertex_t u, v;
+		u = oldG[boost::source(*edgeIt, oldG)].indexOriginal;
+		v = oldG[boost::target(*edgeIt, oldG)].indexOriginal;
+
+		if (vecVertexToComponent[u].find(v) == vecVertexToComponent[u].end() ||
+		    vecVertexToComponent[v].find(u) == vecVertexToComponent[v].end()) {
+			continue;
+		}
+
+		size_t uMolecule = vecVertexToComponent[u][v];
+		size_t vMolecule = vecVertexToComponent[v][u];
+		std::string uName = oldG[u].name + "_" + std::to_string(uMolecule);
+		std::string vName = oldG[v].name + "_" + std::to_string(vMolecule);
+		edge_t e =
+		    boost::add_edge(newGBarcodeToIndex[uName], newGBarcodeToIndex[vName], newG).first;
+		newG[e].weight = oldG[*edgeIt].weight;
+	}
+
+	std::cerr << "Generated new graph ";
+#if _OPENMP
+	std::cerr << "in sec: " << omp_get_wtime() - sTime << std::endl;
+	sTime = omp_get_wtime();
+#endif
+
+	std::cerr << "Memory usage: " << double(memory_usage()) / double(1048576) << "GB" << std::endl;
+}
+
+void
+biconnectedComponents(graph_t& subgraph, vertexToComponent_t& vertexToComponent)
+{
+	// Find biconnected components
+	boost::property_map<graph_t, edgeComponent_t>::type component =
+	    boost::get(edgeComponent, subgraph);
+	boost::biconnected_components(subgraph, component);
+
+	std::vector<vertex_t> artPointsVec;
+	articulation_points(subgraph, std::back_inserter(artPointsVec));
+	std::unordered_set<vertex_t> art_points(artPointsVec.begin(), artPointsVec.end());
```

Comment:
So we call `boost::biconnected_components(subgraph, component)` and then we call `articulation_points(subgraph, std::back_inserter(artPointsVec))`. I am afraid that the bottleneck in computing biconnected components is calculating articulation points and we're calling them twice so it doubles the running time. If it's true, isn't it possible to call only `boost::biconnected_components(subgraph, component)` and do the job based on that?

Reply:
I just figured out that by calling `boost::biconnected_components(subgraph, component)` the components are overlapping by their articulation points, and you need two calls so you remove articulation points from communities to make it consistent with our python implementation.
Although using only `biconnected_components()` make it different from what we had in python but I always wanted to switch to this mode where we do not through out the articluation points themseves.... May produce better results... we can test it later...

Reply:
Not related to this PR, but if it makes sense for you can you please check if muting the `articulation_points` call would reduce the run-time?

Reply:
I had intended to fix this but I forgot. The two functions can be combined